### DISCUSSION TITLE
DB schema: Foreign keys invalid

### DISCUSSION THREAD

**Discussion by mweissdigchg** [dani-garcia/vaultwarden#4651] [Q&A]
### Subject of the issue
Keys and foreign keys mismatch in DB schema (do not have the same datatypes)
This causes issues when initializing DBs with mandatory foreign key checking (like TiDB since 6.6.0) or when you try to restore the database to different databases
mysqldump (mysql-client) marks them with /* FOREIGN KEY INVALID */ (see below)

### Deployment environment

* vaultwarden version: 1.30.5

* Install method: Docker image

* MariaDB version: 11.3.2

### Steps to reproduce
Init new DB by starting fresh instance of vaultwarden on empty DB

### Expected behaviour
Foreign keys and referenced keys have the same data types

### Actual behaviour
as an example:
`collection_uuid` in TABLE `ciphers_collections` has datatype **char(36)** 
FOREIGN KEY (`collection_uuid`) REFERENCES `vaultwarden`.`collections` (`uuid`)
`uuid` in TABLE `collections` has datatype **varchar(40)**

### Troubleshooting data
Check CREATE TABLE statements in database dump, e.g.:

CREATE TABLE `ciphers_collections` (
  `cipher_uuid` char(36) COLLATE utf8mb4_unicode_ci NOT NULL,
  `collection_uuid` char(36) COLLATE utf8mb4_unicode_ci NOT NULL,
  PRIMARY KEY (`cipher_uuid`,`collection_uuid`) /*T![clustered_index] CLUSTERED */,
  KEY `collection_uuid` (`collection_uuid`),
  CONSTRAINT `ciphers_collections_ibfk_1` FOREIGN KEY (`cipher_uuid`) REFERENCES `vaultwarden`.`ciphers` (`uuid`) /* FOREIGN KEY INVALID */,
  CONSTRAINT `ciphers_collections_ibfk_2` FOREIGN KEY (`collection_uuid`) REFERENCES `vaultwarden`.`collections` (`uuid`) /* FOREIGN KEY INVALID */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;

...

CREATE TABLE `collections` (
  `uuid` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `org_uuid` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `name` text COLLATE utf8mb4_unicode_ci NOT NULL,
  `external_id` text COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  PRIMARY KEY (`uuid`) /*T![clustered_index] CLUSTERED */,
  KEY `org_uuid` (`org_uuid`),
  CONSTRAINT `collections_ibfk_1` FOREIGN KEY (`org_uuid`) REFERENCES `vaultwarden`.`organizations` (`uuid`) /* FOREIGN KEY INVALID */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;

**Comment by BlackDex** [comment]
See: https://github.com/dani-garcia/vaultwarden/wiki/Using-the-MariaDB-%28MySQL%29-Backend#foreign-key-errors-collation-and-charset

**Reply by BlackDex** [reply]
You need to define the collation and charset  your self during the creation of the database.
It might be that your backup contained a wrong collation or charset.

I created a new database yesterday for testing using the code currently in the `main` branch. And that works just fine :).

**Reply by mweissdigchg** [reply]
This issue doesn't refer to collation and charset.
Can you please check for example what are the datatypes of "collection_uuid" in TABLE "ciphers_collections" and "uuid" in TABLE "collections" in your database created yesterday. Do they match?
As far as I understand the datatypes are set by vaultwarden during DB initialization.

**Reply by BlackDex** [reply]
All is fine.
And Vaultwarden currently doesn't set any collation or charset, it leaves that to the default settings defined by the user/database.

Query:
```sql
SELECT DEFAULT_CHARACTER_SET_NAME, DEFAULT_COLLATION_NAME FROM information_schema.SCHEMATA WHERE SCHEMA_NAME = "vaultwarden";
```
Output:
|DEFAULT_CHARACTER_SET_NAME|DEFAULT_COLLATION_NAME|
|--------------------------|----------------------|
|utf8mb4|utf8mb4_general_ci|

Query:
```sql
SELECT TABLE_NAME, COLUMN_NAME, CHARACTER_SET_NAME, COLLATION_NAME FROM information_schema.`COLUMNS` WHERE TABLE_SCHEMA = "vaultwarden" AND CHARACTER_SET_NAME IS NOT NULL ORDER BY TABLE_NAME,COLUMN_NAME ASC;
```
Output:
|TABLE_NAME|COLUMN_NAME|CHARACTER_SET_NAME|COLLATION_NAME|
|----------|-----------|------------------|--------------|
|attachments|akey|utf8mb4|utf8mb4_general_ci|
|attachments|cipher_uuid|utf8mb4|utf8mb4_general_ci|
|attachments|file_name|utf8mb4|utf8mb4_general_ci|
|attachments|id|utf8mb4|utf8mb4_general_ci|
|auth_requests|access_code|utf8mb4|utf8mb4_general_ci|
|auth_requests|enc_key|utf8mb4|utf8mb4_general_ci|
|auth_requests|master_password_hash|utf8mb4|utf8mb4_general_ci|
|auth_requests|organization_uuid|utf8mb4|utf8mb4_general_ci|
|auth_requests|public_key|utf8mb4|utf8mb4_general_ci|
|auth_requests|request_device_identifier|utf8mb4|utf8mb4_general_ci|
|auth_requests|request_ip|utf8mb4|utf8mb4_general_ci|
|auth_requests|response_device_id|utf8mb4|utf8mb4_general_ci|
|auth_requests|user_uuid|utf8mb4|utf8mb4_general_ci|
|auth_requests|uuid|utf8mb4|utf8mb4_general_ci|
|ciphers|data|utf8mb4|utf8mb4_general_ci|
|ciphers|fields|utf8mb4|utf8mb4_general_ci|
|ciphers|key|utf8mb4|utf8mb4_general_ci|
|ciphers|name|utf8mb4|utf8mb4_general_ci|
|ciphers|notes|utf8mb4|utf8mb4_general_ci|
|ciphers|organization_uuid|utf8mb4|utf8mb4_general_ci|
|ciphers|password_history|utf8mb4|utf8mb4_general_ci|
|ciphers|user_uuid|utf8mb4|utf8mb4_general_ci|
|ciphers|uuid|utf8mb4|utf8mb4_general_ci|
|ciphers_collections|cipher_uuid|utf8mb4|utf8mb4_general_ci|
|**ciphers_collections**|**collection_uuid**|**utf8mb4**|**utf8mb4_general_ci**|
|collections|external_id|utf8mb4|utf8mb4_general_ci|
|collections|name|utf8mb4|utf8mb4_general_ci|
|collections|org_uuid|utf8mb4|utf8mb4_general_ci|
|**collections**|**uuid**|**utf8mb4**|**utf8mb4_general_ci**|
|collections_groups|collections_uuid|utf8mb4|utf8mb4_general_ci|
|collections_groups|groups_uuid|utf8mb4|utf8mb4_general_ci|
|devices|name|utf8mb4|utf8mb4_general_ci|
|devices|push_token|utf8mb4|utf8mb4_general_ci|
|devices|push_uuid|utf8mb4|utf8mb4_general_ci|
|devices|refresh_token|utf8mb4|utf8mb4_general_ci|
|devices|twofactor_remember|utf8mb4|utf8mb4_general_ci|
|devices|user_uuid|utf8mb4|utf8mb4_general_ci|
|devices|uuid|utf8mb4|utf8mb4_general_ci|
|emergency_access|email|utf8mb4|utf8mb4_general_ci|
|emergency_access|grantee_uuid|utf8mb4|utf8mb4_general_ci|
|emergency_access|grantor_uuid|utf8mb4|utf8mb4_general_ci|
|emergency_access|key_encrypted|utf8mb4|utf8mb4_general_ci|
|emergency_access|uuid|utf8mb4|utf8mb4_general_ci|
|event|act_user_uuid|utf8mb4|utf8mb4_general_ci|
|event|cipher_uuid|utf8mb4|utf8mb4_general_ci|
|event|collection_uuid|utf8mb4|utf8mb4_general_ci|
|event|group_uuid|utf8mb4|utf8mb4_general_ci|
|event|ip_address|utf8mb4|utf8mb4_general_ci|
|event|org_user_uuid|utf8mb4|utf8mb4_general_ci|
|event|org_uuid|utf8mb4|utf8mb4_general_ci|
|event|policy_uuid|utf8mb4|utf8mb4_general_ci|
|event|provider_org_uuid|utf8mb4|utf8mb4_general_ci|
|event|provider_user_uuid|utf8mb4|utf8mb4_general_ci|
|event|provider_uuid|utf8mb4|utf8mb4_general_ci|
|event|user_uuid|utf8mb4|utf8mb4_general_ci|
|event|uuid|utf8mb4|utf8mb4_general_ci|
|favorites|cipher_uuid|utf8mb4|utf8mb4_general_ci|
|favorites|user_uuid|utf8mb4|utf8mb4_general_ci|
|folders|name|utf8mb4|utf8mb4_general_ci|
|folders|user_uuid|utf8mb4|utf8mb4_general_ci|
|folders|uuid|utf8mb4|utf8mb4_general_ci|
|folders_ciphers|cipher_uuid|utf8mb4|utf8mb4_general_ci|
|folders_ciphers|folder_uuid|utf8mb4|utf8mb4_general_ci|
|groups|external_id|utf8mb4|utf8mb4_general_ci|
|groups|name|utf8mb4|utf8mb4_general_ci|
|groups|organizations_uuid|utf8mb4|utf8mb4_general_ci|
|groups|uuid|utf8mb4|utf8mb4_general_ci|
|groups_users|groups_uuid|utf8mb4|utf8mb4_general_ci|
|groups_users|users_organizations_uuid|utf8mb4|utf8mb4_general_ci|
|invitations|email|utf8mb4|utf8mb4_general_ci|
|organizations|billing_email|utf8mb4|utf8mb4_general_ci|
|organizations|name|utf8mb4|utf8mb4_general_ci|
|organizations|private_key|utf8mb4|utf8mb4_general_ci|
|organizations|public_key|utf8mb4|utf8mb4_general_ci|
|organizations|uuid|utf8mb4|utf8mb4_general_ci|
|organization_api_key|api_key|utf8mb4|utf8mb4_general_ci|
|organization_api_key|org_uuid|utf8mb4|utf8mb4_general_ci|
|organization_api_key|uuid|utf8mb4|utf8mb4_general_ci|
|org_policies|data|utf8mb4|utf8mb4_general_ci|
|org_policies|org_uuid|utf8mb4|utf8mb4_general_ci|
|org_policies|uuid|utf8mb4|utf8mb4_general_ci|
|sends|akey|utf8mb4|utf8mb4_general_ci|
|sends|data|utf8mb4|utf8mb4_general_ci|
|sends|name|utf8mb4|utf8mb4_general_ci|
|sends|notes|utf8mb4|utf8mb4_general_ci|
|sends|organization_uuid|utf8mb4|utf8mb4_general_ci|
|sends|user_uuid|utf8mb4|utf8mb4_general_ci|
|sends|uuid|utf8mb4|utf8mb4_general_ci|
|twofactor|data|utf8mb4|utf8mb4_general_ci|
|twofactor|user_uuid|utf8mb4|utf8mb4_general_ci|
|twofactor|uuid|utf8mb4|utf8mb4_general_ci|
|twofactor_incomplete|device_name|utf8mb4|utf8mb4_general_ci|
|twofactor_incomplete|device_uuid|utf8mb4|utf8mb4_general_ci|
|twofactor_incomplete|ip_address|utf8mb4|utf8mb4_general_ci|
|twofactor_incomplete|user_uuid|utf8mb4|utf8mb4_general_ci|
|users|akey|utf8mb4|utf8mb4_general_ci|
|users|api_key|utf8mb4|utf8mb4_general_ci|
|users|avatar_color|utf8mb4|utf8mb4_general_ci|
|users|email|utf8mb4|utf8mb4_general_ci|
|users|email_new|utf8mb4|utf8mb4_general_ci|
|users|email_new_token|utf8mb4|utf8mb4_general_ci|
|users|equivalent_domains|utf8mb4|utf8mb4_general_ci|
|users|excluded_globals|utf8mb4|utf8mb4_general_ci|
|users|external_id|utf8mb4|utf8mb4_general_ci|
|users|name|utf8mb4|utf8mb4_general_ci|
|users|password_hint|utf8mb4|utf8mb4_general_ci|
|users|private_key|utf8mb4|utf8mb4_general_ci|
|users|public_key|utf8mb4|utf8mb4_general_ci|
|users|security_stamp|utf8mb4|utf8mb4_general_ci|
|users|stamp_exception|utf8mb4|utf8mb4_general_ci|
|users|totp_recover|utf8mb4|utf8mb4_general_ci|
|users|totp_secret|utf8mb4|utf8mb4_general_ci|
|users|uuid|utf8mb4|utf8mb4_general_ci|
|users_collections|collection_uuid|utf8mb4|utf8mb4_general_ci|
|users_collections|user_uuid|utf8mb4|utf8mb4_general_ci|
|users_organizations|akey|utf8mb4|utf8mb4_general_ci|
|users_organizations|external_id|utf8mb4|utf8mb4_general_ci|
|users_organizations|org_uuid|utf8mb4|utf8mb4_general_ci|
|users_organizations|reset_password_key|utf8mb4|utf8mb4_general_ci|
|users_organizations|user_uuid|utf8mb4|utf8mb4_general_ci|
|users_organizations|uuid|utf8mb4|utf8mb4_general_ci|
|__diesel_schema_migrations|version|utf8mb4|utf8mb4_general_ci|

**Reply by mweissdigchg** [reply]
Again, I am not refering to collation and charset. To see the datatypes and max length check out this SQL query:
```
SELECT TABLE_NAME, COLUMN_NAME, DATA_TYPE, CHARACTER_MAXIMUM_LENGTH
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_NAME = 'collections' or TABLE_NAME = 'ciphers_collections'
```

**Reply by BlackDex** [reply]
That shows the following.

While that is probably not optimal, it shouldn't cause major issues. This is there for more then 5 years already. Not saying that it is correct, but there were no major issues before regarding this.

We might want to update all those kind of fields to match all over the whole code-base. But not sure what impact that will have right now.

|TABLE_NAME|COLUMN_NAME|DATA_TYPE|CHARACTER_MAXIMUM_LENGTH|
|----------|-----------|---------|------------------------|
|collections|uuid|varchar|40|
|collections|org_uuid|varchar|40|
|collections|name|text|65535|
|collections|external_id|text|65535|
|ciphers_collections|cipher_uuid|char|36|
|ciphers_collections|collection_uuid|char|36|

**Reply by BlackDex** [reply]
Also during my `mysqldump` (a.k.a. `mariadb-dump`) it doesn't report any foreign-key as invalid.
And i can export and import without any issues based upon the dump.

**Reply by mweissdigchg** [reply]
OK, I will redo my dump as the above was done on another machine a few weeks ago. I think I had mysql-client on that machine.

That's what I found in the DB docs so far:

> The foreign key columns and the referenced columns must be of the same type, or similar types. For integer types, the size and sign must also be the same

Source: https://mariadb.com/kb/en/foreign-keys/

> Corresponding columns in the foreign key and the referenced key must have similar data types.

Source: https://dev.mysql.com/doc/refman/8.4/en/create-table-foreign-keys.html#foreign-key-restrictions

**Reply by BlackDex** [reply]
They are the same type, it's char/varchar/text, only the size is different, which according to that statement is only an issue for integers.

**Reply by mweissdigchg** [reply]
You are right. It seems my dump was from mysql-compatible TiDB. I have to find out why only here it is marked as `/* FOREIGN KEY INVALID */`

You were able to import because foreign key checks is disabled before import with the following command, otherwise you wouldn't be able to create the tables:
`/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;`

**Reply by mweissdigchg** [reply]
> They are the same type, it's char/varchar/text, only the size is different, which according to that statement is only an issue for integers.

I don't think they are the same, but I am not an SQL expert. I will try to catch one the next days or possibly community knows more on this topic?

**Reply by BlackDex** [reply]
They are, see https://dev.mysql.com/doc/refman/8.4/en/string-types.html
